app/models/cross_model_slug.rb

```diff
@@ -0,0 +1,42 @@
+#
+# We have a simple top-level route equivalent to `/slug`. Because of this,
+# we want to verify that newly created records don't overlap with a previously-
+# defined `slug` -- in other words, slug values should be unique across all
+# relevant models. Except also, models might use "username" instead of "slug".
+#
+# Models can register with this class to be included in a cross-model-uniqueness
+# check. Any/all models registered are checked for the existence of a record
+# matching a given value. Additionally, we have some special cases (eg, sitemap)
+# that we want to apply across all registered models.
+#
+class CrossModelSlug
+  class << self
+    attr_accessor :registered_models
+
+    def register(klass, attribute)
+      @registered_models ||= []
+      @registered_models << [klass, attribute]
+    end
+
+    # This is currently equivalent to:
+    # User.exists?(username: username) ||
+    #    Organization.exists?(slug: username) ||
+    #    Page.exists?(slug: username) ||
+    #    Podcast.exists?(slug: username)
+    def exists?(value)
+      # Presence check is likely redundant, but is **much** cheaper than the
+      # cross-model check
+      return false if value.blank?
+
+      value = value.downcase
+
+      # Reserved check may be redundant, but allows this to be used outside of Validator
+      return true if ReservedWords.all.include?(value)
+      return true if value.include?("sitemap-") # https://github.com/forem/forem/pull/6704
+
+      (@registered_models || []).detect do |klass, attribute|
+        klass.exists?({ attribute => value })
+      end
```

Comment:
I understand that `@registered_models` should have all proper model loaded on boot. Out of curiosity, I tried to check `CrossModelSlug.registered_models` in rails console, but only User model is registered. This got me thinking, would there be a scenario where not all the models were properly registered? I'm thinking in the case of background jobs.

Reply:
So, any `production` environment ought to have `eager_load = true`, which will cause Rails to parse every file in the `eager_load_paths`. In `development` and `test`, `eager_load` is `false`, because Rails prefers to refresh code on every request in development.

```
# with eager-loading **off** / development & test
❯ rails r 'puts CrossModelSlug.registered_models.map(&:first).map(&:to_s)'
[...warnings...]
User

# with eager-loading **on** / production
❯ rails r 'puts CrossModelSlug.registered_models.map(&:first).map(&:to_s)'
[...warnings...]
User
Organization
Page
Podcast
```

(It's somewhat interesting that `User` is already loaded even in development, the implication is that something in the Rails env boot cycle — a gem, an initializer, or routes maybe — references `User` somehow.)

This isn't to say that this is 0% risk:

* We might introduce new environments that aren't `eager_load` (maybe a `staging` environment? or a `asset-build` environment?)
* We might trigger a kind of "false negative" while testing or while QAing in `development` — this issue originally came up while [working on API changes](https://github.com/forem/forem/pull/19009/files/dff9352872f95a4ef17c07c6996d1ad82c538878#diff-92779dbf72cdd79815877e327b5693b0448053945e302d567814c33a5e0a49d1R14), when I noticed I could create multiple Pages with the same slug, that kind of confusion could come up again

(Traditionally, the biggest risk with this kind of class-level macro is creating a "loop" — the interpreter can't load `User` without loading some `Module` that then references `User`, which then references `Module`, ➿ . The "lazy" effect here — registration, using that registration at a later point — does not have _that_ problem.)

Just to be clear: I am a bit conflicted on this `registration` approach — I tend to think the trade-offs are "worth it", but even I think this is dancing very close to "too much".

Reply:
Hmm, well, the potential for risk here is also giving me a bit of hesitancy. I like clever approaches, but it may be worth just sticking toe the verbose explicit 'registration' to have that peace of mind. I would like to hear others opinion on this.

Reply:
I'm not comfortable with "registering", as it brings the risks, I would set up the explicit list of the models, across which the validator works, and so, needs to be checked.

Reply:
Removed registration in [82cbb121](https://github.com/forem/forem/pull/19263/commits/82cbb1215c74949714094e10547cf464e302ed35).